### PR TITLE
Central implementation of syntactic equality of instructions and goto programs

### PR DESCRIPTION
**Pull request by tautschnig**

### CODE REVIEW

**Comment by tautschnig** [comment]
Factored out from #1987 and now includes the mirrored changes to java_syntactic_diff.cpp.

File: src/goto-programs/goto_program.cpp [hunk 1]
```diff
@@ -634,3 +634,45 @@ void goto_programt::compute_incoming_edges()
...
+bool goto_programt::instructiont::equals(const instructiont &other) const
+{
+  // clang-format off
+  return
+    type == other.type &&
+    code == other.code &&
+    guard == other.guard &&
+    targets.size() == other.targets.size() &&
+    labels == other.labels;
+  // clang-format on
```
Comment (kroening) [inline src/goto-programs/goto_program.cpp:646]: This might be confusing as the target doesn't get compared?

Comment (tautschnig) [reply src/goto-programs/goto_program.cpp:646]: In my opinion that's not an aspect of equality that can be resolved on a single instruction. That's contextual information that only makes sense when looking at the program, I'd say.

Comment (tautschnig) [reply src/goto-programs/goto_program.cpp:646]: On a second thought: maybe the name of the function should be changed to convey this? And suggestions?

Comment (peterschrammel) [reply src/goto-programs/goto_program.cpp:646]: Where is this function intended to be used? 
The implementation that we use in goto-diff compares the target offset at least.
It should at least be true that if all instructions of a function are the same then the function is the same. Or do you want to separate equality of 'bare' instructions from comparing the control flow?

Comment (tautschnig) [reply src/goto-programs/goto_program.cpp:646]: See the second commit in this PR. Edit: it's even in this commit already, line 657.

**Review by peterschrammel — APPROVED** [review]

**Comment by peterschrammel** [comment]
I think this is good to go.

**Pull request merged**